pkg/trie/triedb/codec/node.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package codec
+
+const ChildrenCapacity = 16
+
+// MerkleValue is a helper enum to differentiate between inline and hashed nodes
+// https://spec.polkadot.network/chap-state#defn-merkle-value
+type MerkleValue interface {
+	_isMerkleValue()
+	IsHashed() bool
+}
```

Comment:
Instead of having a `MerkleValue` what about introducing a new item in to the Node variant called ref Ref

```go
type (
	// Empty node
	// Leaf always contains values
	// Branch could has or not has values
        // Ref is either a hash for a node or the node itself if Data is less than 32 bytes
	Ref struct {
		Data []byte
	}
)
```

So when you're decoding the branch children you always will have a 

```go
node.Children[i] = NewNodeRef(...)
```

which makes easy for looking up also, given that you will have a variant for node references, then you only need to check if the ref size and decode accordingly

Reply:
But we are going to need to check the size to know if it is hashed or inlined... mmm I got your point but I'm not sure if this is an improvement

Reply:
I just think that is easy to see them as references instead of merkle values, I would say that the `MerkleValue` and `NodeValue` seems confusing to me given that one refers to a byte slice that can become a node (`MerkleValue`) and the other is just the value contained by a node. So I still think that you can treat a byte slice that can be decoded into a node as a `Node::Ref`, and is responsibility of the decoder to check the size in order to decode directly or search in the database for the actual node bytes, if think that is better to have 2 variants for inline and hashed (`Node::InlineRef` and a `Node::HashedRef`) is up to you but I think these values should belong to the Node enum

Reply:
I'm still thinking that it is better to differentiate this "special nodes" using the `MerkleValue` structure.

- One reason is that it is the same concept in the [spec](https://spec.polkadot.network/chap-state#defn-merkle-value)
- Another reason is that if we change them, since they are "special nodes" we cannot get the partial key or the value from them since they are not decoded yet, so that will force us to add methods with empty bodies if we want to add more methods to the interface as I'm doing [here](https://github.com/ChainSafe/gossamer/pull/3884/files#diff-26147d32b2f68223030e0024afd41fc96e346f0b7fcdf14e065688df4d3c3248)

Don't be mad with me but I'll keep my vote on leaving them as they are now 😅

Reply:
I will not be mad hahah! Okay, there is no problem at all and we can progress with your impl, but I would like to add some thoughts: 
- the first reason does not convince me too much since spec is more a general way of describing the idea and not much about how we should implement.
- For the second point: yeah, they are nodes but described in another form but they still belong to the nodes group, so for that reason we can fit them in the `Node` interface. 
> we cannot get the partial key or the value from them since they are not decoded yet

so it is like:
```go 
var a *int
fmt.Println(a)
```
we cannot retrieve the value that `a` holds, to that we need to deref the pointer, to do so we need access to the heap where the data is, so I think we can have the same to the `Ref` type, it is responsability to the `Ref` to retrieve the data when needed, like deref is resposability of the pointer in order to bring the actual value

```go
type Node interface {
   paritalKey()
   value()
}

type Ref struct {
  db DatabaseGetter
  encoded []byte
}

fn (r *Ref) partialKey() {
  _, keyLen := decodeHeader(r.encoded)
  return decodeKey(r.encoded, keyLen)
}

fn (r *Ref) value() {
 reader := bytes.NewReader(r.encoded)
 nodeKind, _ := decodeHeader(reader)
 // you might need the database if the value is just
 // a hash to a value store in database
 return decodeNodeValue(reader, nodeKind, r.db)
}
```

Reply:
I agree that it is feasible, but that's not the main blocker. 
Implementing this change would require us to modify the `Decode` function to add the `db` as part of the node. Currently, the `codec` is designed without any knowledge of the `db`. 
My goal with these new changes is to maintain the code's simplicity and avoid unnecessary complexity, in line with the [KISS](https://en.wikipedia.org/wiki/KISS_principle) principle. Therefore, although the proposed change might seem sensible, I believe we should defer it and possibly reconsider in the future if there's a compelling need.

Let me know if you are ok with this so we can resolve this conversation. 
Thanks you again for the discussion ❤️